src/coffee/product-import.coffee

```diff
@@ -7,5 +9,131 @@ class ProductImport
   constructor: (@logger, options = {}) ->
     @sync = new ProductSync
     @client = new SphereClient options
+    @_resetCache()
+    @_resetSummary()
+
+  _resetCache: ->
+    @_cache =
+      productType: {}
+      categories: {}
+      taxCategory: {}
+
+  _resetSummary: ->
+    @_summary =
+      emptySKU: 0
+      created: 0
+      updated: 0
+
+
+  performStream: (chunk, cb) ->
+    @_processBatches(chunk).then -> cb()
+
+  _processBatches: (products) ->
+    batchedList = _.batchList(products, 30) # max parallel elem to process
+    Promise.map batchedList, (productsToProcess) =>
+      # extract all skus from master variant and variants of all jsons in the batch
+      skus = @_extractUniqueSkus(productsToProcess)
+      predicate = @_prepareProductFetchBySkuQueryPredicate(skus)
+      # Check predicate size by: Buffer.byteLength(predicate,'utf-8')
+      # Todo: Handle predicate if predicate size > 8kb
+      # Fetch products from product projections end point by list of skus.
+      @client.productProjections
+      .where(predicate)
+      .fetch()
+      .then (results) =>
+        debug 'Fetched products: %j', results
+        queriedEntries = results.body.results
+        @_createOrUpdate productsToProcess, queriedEntries
+      .then (results) =>
+        _.each results, (r) =>
+          switch r.statusCode
+            when 201 then @_summary.created++
+            when 200 then @_summary.updated++
+        Promise.resolve()
+    ,{concurrency: 1} # run 1 batch at a time
+
+
+  _prepareProductFetchBySkuQueryPredicate: (skus) ->
+    skuString = "sku in (\"#{skus.join('", "')}\")"
+    return "masterVariant(#{skuString}) or variants(#{skuString})"
+
+  _extractUniqueSkus: (products) ->
+    skus = []
+    for product in products
```

Comment:
Actually I think it's ok like this, since we are also selecting only valid data.